src/test/java/rx/internal/util/ScalarSynchronousSingleTest.java

```diff
@@ -0,0 +1,285 @@
+
+public class ScalarSynchronousSingleTest {
+    @Test
+    public void testBackpressure() {
+        TestSubscriber<Integer> ts = TestSubscriber.create(0);
+
+        Single.just(1).subscribe(ts);
+
+        ts.assertNoValues();
+        ts.assertNoErrors();
+        ts.assertNotCompleted();
+
+        ts.requestMore(1);
+
+        ts.assertValue(1);
+        ts.assertCompleted();
+        ts.assertNoErrors();
+
+        ts.requestMore(1);
+
+        ts.assertValue(1);
+        ts.assertCompleted();
+        ts.assertNoErrors();
+    }
+
+    @Test(timeout = 1000)
+    public void testBackpressureSubscribeOn() throws Exception {
+        TestSubscriber<Integer> ts = TestSubscriber.create(0);
+
+        Single.just(1).subscribeOn(Schedulers.computation()).subscribe(ts);
+
+        Thread.sleep(200);
+
+        ts.assertNoValues();
+        ts.assertNoErrors();
+        ts.assertNotCompleted();
+
+        ts.requestMore(1);
+
+        ts.awaitTerminalEvent();
+
+        ts.assertValue(1);
+        ts.assertCompleted();
+        ts.assertNoErrors();
+    }
+
+    @Test(timeout = 1000)
+    public void testBackpressureObserveOn() throws Exception {
+        TestSubscriber<Integer> ts = TestSubscriber.create(0);
+
+        Single.just(1).observeOn(Schedulers.computation()).subscribe(ts);
+
+        Thread.sleep(200);
+
+        ts.assertNoValues();
+        ts.assertNoErrors();
+        ts.assertNotCompleted();
+
+        ts.requestMore(1);
+
+        ts.awaitTerminalEvent();
+
+        ts.assertValue(1);
+        ts.assertCompleted();
+        ts.assertNoErrors();
+    }
+
+    @Test(timeout = 1000)
+    public void testBackpressureSubscribeOn2() throws Exception {
+        TestSubscriber<Integer> ts = TestSubscriber.create(0);
+
+        Single.just(1).subscribeOn(Schedulers.newThread()).subscribe(ts);
+
+        Thread.sleep(200);
+
+        ts.assertNoValues();
+        ts.assertNoErrors();
+        ts.assertNotCompleted();
+
+        ts.requestMore(1);
+
+        ts.awaitTerminalEvent();
+
+        ts.assertValue(1);
+        ts.assertCompleted();
+        ts.assertNoErrors();
+    }
+
+    @Test(timeout = 1000)
+    public void testBackpressureObserveOn2() throws Exception {
+        TestSubscriber<Integer> ts = TestSubscriber.create(0);
+
+        Single.just(1).observeOn(Schedulers.newThread()).subscribe(ts);
+
+        Thread.sleep(200);
+
+        ts.assertNoValues();
+        ts.assertNoErrors();
+        ts.assertNotCompleted();
+
+        ts.requestMore(1);
+
+        ts.awaitTerminalEvent();
+
+        ts.assertValue(1);
+        ts.assertCompleted();
+        ts.assertNoErrors();
+    }
+
+    @Test
+    public void testBackpressureFlatMapJust() {
+        TestSubscriber<Integer> ts = TestSubscriber.create(0);
+
+        Single.just(1).flatMap(new Func1<Integer, Single<Integer>>() {
+            @Override
+            public Single<Integer> call(Integer v) {
+                return Single.just(v);
```

Comment:
Please return another result, at the moment test can pass even if `flatMap` returns same `Single`, you can also change type from `Integer` to let's say `String` during `flatMap` :)